pkg/cluster/install.go

```diff
@@ -141,54 +142,100 @@ func (m *manager) Update(ctx context.Context) error {
+	// code since it's easier, but in the future, this data should be collected
+	// from Hive's outputs where needed.
+	return m.hiveClusterManager.Install(ctx, m.subscriptionDoc, m.doc, v)
+}
+
+func (m *manager) bootstrap() []steps.Step {
+	s := []steps.Step{
+		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.validateResources)),
+		steps.Action(m.ensureACRToken),
+		steps.Action(m.ensureInfraID),
+		steps.Action(m.ensureSSHKey),
+		steps.Action(m.ensureStorageSuffix),
+		steps.Action(m.populateMTUSize),
+
+		steps.Action(m.createDNS),
+		steps.Action(m.initializeClusterSPClients), // must run before clusterSPObjectID
+		steps.Action(m.clusterSPObjectID),
+		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.ensureResourceGroup)),
+		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.enableServiceEndpoints)),
+		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.setMasterSubnetPolicies)),
+		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.deployStorageTemplate)),
+		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.attachNSGs)),
+		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.updateAPIIPEarly)),
+		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.createOrUpdateRouterIPEarly)),
+		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.ensureGatewayCreate)),
+		steps.Action(m.createAPIServerPrivateEndpoint),
+		steps.Action(m.createCertificates),
+
+		// We will always need a Hive namespace, whether we are installing
+		// via Hive or adopting
+		steps.Action(m.hiveCreateNamespace),
+	}
+
+	if m.installViaHive {
+		s = append(s,
+			steps.Action(m.hiveCreateNamespace),
+			steps.Action(m.runHiveInstaller),
+			// Give Hive 60 minutes to install the cluster, since this includes
+			// all of bootstrapping being complete
+			steps.Condition(m.hiveClusterInstallationComplete, 60*time.Minute, true),
+			steps.Condition(m.hiveClusterDeploymentReady, 5*time.Minute, true),
+		)
+	} else {
+		s = append(s,
+			steps.Action(m.runIntegratedInstaller),
 			steps.Action(m.hiveCreateNamespace),
 			steps.Action(m.hiveEnsureResources),
 			steps.Condition(m.hiveClusterDeploymentReady, 5*time.Minute, true),
-			steps.Action(m.hiveResetCorrelationData),
+		)
+	}
 
-			steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.generateKubeconfigs)),
-			steps.Action(m.ensureBillingRecord),
-			steps.Action(m.initializeKubernetesClients),
-			steps.Action(m.initializeOperatorDeployer), // depends on kube clients
-			steps.Condition(m.apiServersReady, 30*time.Minute, true),
-			steps.Action(m.ensureAROOperator),
-			steps.Action(m.incrInstallPhase),
-		},
+	s = append(s,
+		// Reset correlation data whether adopting or installing via Hive
+		steps.Action(m.hiveResetCorrelationData),
+
+		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.generateKubeconfigs)),
```

Comment:
There is a bug here for hive adoption, in that `m.hiveEnsureResources` runs before `m.generateKubeconfigs` and therefore the kubeconfig sent to Hive is empty. I have fixed this in my follow up PR, so let's not worry about it here either.